OpenSim/Simulation/Model/Model.cpp

```diff
@@ -404,6 +404,16 @@ void Model::buildSystem() {
     // necessary elements to the System. Doesn't initialize geometry yet.
     if (getUseVisualizer())
         _modelViz.reset(new ModelVisualizer(*this));
+
+    // Cache enabled controllers locally so that `computeControls` can
+    // iterate over the cache, rather than needing to call
+    // `getComponentList` (expensive) on each iteration.
+    this->_enabledControllers.clear();
+    for (const Controller& controller : getComponentList<Controller>()) {
+        if (controller.isEnabled()) {
+            this->_enabledControllers.emplace_back(controller);
+        }
+    }
```

Comment:
Why here instead of Model's `extendFinalizeConnections()` ?

Reply:
Genuinely, I didn't know. I'll move it in there if you think that's the best place.

This is a good opportunity to point out how an experienced, but generic, developer might look at this task. My goal is to cache some data just before computation happens. So i'm fixated on finding out where in the lifecycle I should initialize the cache. 

One barrier to understanding OpenSim is that it uses inheritence, rather than interface implementation, to achieve its functionality. This means that I need to traverse the entire inheritence hierarchy in order to understand what data structures are touched and when, rather than the entire implementation being in one place.

The other barrier is that the lifecycle methods aren't automatically obvious to me. I'm expecting `beforeComponentInitialization`, `beforeComponentConnection`, etc. etc. but there's `finalizeFromX`-type methods, which imply that `Model` will be "finalized" after that's called. 

Here's a brief tour of my mind, reading through the classes. Note: I, like every other developer, am just looking to get my fix working without carefully reading through the >150 methods that `Model+ModelComponent+Component+Object`, so I'm going to stop as soon as something sounds reasonable and works (e.g. when I see `buildSystem`):

```c++
// Oh boy, this system spends *a lot* of time getting and downcasting things.
// Lets put a bandaid on that by adding a cache, rather than specializing the
// storage inside `Model` itself.

// Sounds like what I need, but isn't: it's "some setup functions that are called after
// it's been deserialized" and it's broken on copies
void Model::setup();

// Could be here, because I don't *really* need the Model::extendFinalizeConnections() 
// trampoline, but it shouldn't go here because callers *might* use 
// `Component::finalizeConnections(Component&)`
void Model::finalizeConnections();

// Seems reasonable: I want to freeze the cache after building the SimTK::System.
void Model::buildSystem();

// Also could work, because this method is definitely called when the connectivity is
// finalized and the caller wants to create a state
SimTK::State& Model::initializeState();

// Could work here, but it's just a sugar method and there's no guarantee a caller
// will call this sugar rather than each substep separately.
SimTK::State& Model::initSystem();

// Sounds like a footgun
void Model::initStateWithoutRecreatingSystem(SimTK::State& state);

// hm, alright, going to pretend "Controls" is something that's immutable and hope to god
// this doesn't invalidate my caching later down the line *whistles ominously*
ControllerSet& Model::updControllerSet();

// seems pertinent, but I think this flag was ignored in the original implementation, 
// I mayaswell ignore it also
void Model::setAllControllersEnabled( bool enabled );

// same deal with any other mutators
void Model::addController(Controller *adoptee);

// maybe?
void Model::extendFinalizeFromProperties();

// maybe?
void Model::createMultibodyTree();
void Model::createMultibodySystem();

// wtf? A base class needs private access to a derived class? Ignoring
friend class Component;

// Recurse to `ModelComponent`

// Something about satisfying `Component` and how it's going to handle making sure
// the method is called on **subcomponents**, maybe? I want *my* cache to be initialized
// not subcomponents'
void ModelComponent::extendFinalizeConnections(Component& root) override final;

// Recurse to `Component`

// extremely long comment string for doucmentation. Doesn't seem to mention initialization
// or method lifecycles. Seems to re-state how SimTK works: ignore

// seems, and sounds, like a winner, also explicitly mentions that it's
// a depth-first traversal followed by calling on itself.
void Component::finalizeFromProperties();

// hm, ok, but isn't it already finalized if I `__finalize__fromProperties`?
// I'm guessing there's finalizing from properties when the properties are
// loaded from XML and there's finalizing connections, which might be called
// when the model is made in-memory? Non-virtual, so ignoring.
void Component::finalizeConnections(Component& root);

// maybe?
void Component::initStateFromProperties(SimTK::State& state) const;

// this is the bastard that should be abstractly defined so that `Model` can,
// itself, handle storage and specialize it by type
ComponentList<T> Component::getComponentList();
// loads of other getters and iterators that downcast things because 
// subclasses don't define the storage model

// 58 other, mostly inlined, methods. Something to do with cache variables, 
// which I'm hoping refers to SimTK  cache variables, rather than some other
// OpenSim caching I need to keep in-mind

// L1785: oooh, this might be something! Ah, and it mentions that it's a "not yet
// connected" model. Maybe I sort out the caching here?
virtual void Component::extendFinalizeFromProperties() {};

// L1810: --------------what you want--------------------
// seems to mention setting up a DAG of components, maybe I want to cache
// *after* the DAG is set up, sure, but my goal (again) is to initialize a cache *just*
// before computation happens. It's not clear where in the lifecycle this is because it
// mentions:
//
// - prior to creating the Simbody MultibodySystem to represent it computationally
// - It may also be invoked at times just for its error-checking side effects
virtual void Component::extendFinalizeConnections(Component& root) {};

// ~60 other protected/public methods: might be in there

















```

Reply:
It is complicated. `buildSystem` works, but I just wanted to point you to the connections phase of preparing the model.  I am happy to help sift through the weeds.